modules/channel-messenger/src/backend/messenger.ts

```diff
@@ -0,0 +1,253 @@
+import axios, { AxiosInstance } from 'axios'
+import bodyParser from 'body-parser'
+import * as sdk from 'botpress/sdk'
+import crypto from 'crypto'
+import { Router } from 'express'
+import _ from 'lodash'
+
+import { Config } from '../config'
+
+const outgoingTypes = ['text', 'typing', 'login_prompt', 'carousel']
+type MessengerAction = 'typing_on' | 'typing_off' | 'mark_seen'
+
+export class MessengerService {
+  private readonly http = axios.create({ baseURL: 'https://graph.facebook.com/v2.6/me' })
+
+  private messengerClients: { [key: string]: MessengerClient } = {}
+  private router: Router
+
+  constructor(private bp: typeof sdk) {}
+
+  initialize() {
+    this.router = this.bp.http.createRouterForBot('channel-messenger', { checkAuthentication: false })
+    this.router.use(
+      bodyParser.json({
+        verify: this._verifySignature.bind(this)
+      })
+    )
+
+    this.router.get('/webhook', this._setupWebhook.bind(this))
+    this.router.post('/webhook', this._handleIncomingMessage.bind(this))
+
+    this.bp.events.registerMiddleware({
+      description: 'Sends outgoing messages for the messenger channel',
+      direction: 'outgoing',
+      handler: this._handleOutgoingEvent.bind(this),
+      name: 'messenger.sendMessages',
+      order: 200
+    })
+  }
+
+  getMessengerClient(botId: string): MessengerClient {
+    if (this.messengerClients[botId]) {
+      return this.messengerClients[botId]
+    }
+
+    this.messengerClients[botId] = new MessengerClient(botId, this.bp, this.http)
+    return this.messengerClients[botId]
+  }
+
+  // See: https://developers.facebook.com/docs/messenger-platform/webhook#security
+  private async _verifySignature(req, res, buffer) {
+    const client = this.getMessengerClient(req.params.botId)
+    const config = await client.getConfig()
+    const signatureError = new Error("Couldn't validate the request signature.")
+
+    if (!/^\/webhook/i.test(req.path)) {
```

Comment:
this check is a little useless since you are the one who calls this verify function in the /webhook route and no where else. This will then always be true except if you change the /webhook route to something else then you will have to manually migrate this check to the new route name .. 

It could also be false if you add a route to your router since you use this middleware at the router level, then I suggest you to move it to route level since you want it to be route scoped

Reply:
facebook requires the `/webhook` endpoint. Furthermore this also ensure that the request comes from facebook and not from somewhere else.